File: assignment_09_simple_calculator.py
# =============================================================================
# PROGRAMMING FUNDAMENTALS — Assignment 9
# =============================================================================
#
# TASK: Console-Based Simple Calculator
#
# Build a calculator program that runs in the console and performs basic
# arithmetic operations based on the user's input.
#
# -----------------------------------------------------------------------------
# OPERATIONS YOUR CALCULATOR MUST SUPPORT
# -----------------------------------------------------------------------------
#
#   1. Addition          ( + )    e.g.  10 + 3  =  13
#   2. Subtraction       ( - )    e.g.  10 - 3  =  7
#   3. Multiplication    ( * )    e.g.  10 * 3  =  30
#   4. Division          ( / )    e.g.  10 / 3  =  3.33
#   5. Modulus           ( % )    e.g.  10 % 3  =  1  (remainder)
#   6. Exponentiation    ( ** )   e.g.  2 ** 8  =  256
#   7. Quit
#
# -----------------------------------------------------------------------------
# HOW THE MENU SHOULD LOOK
# -----------------------------------------------------------------------------
#
#   ============================
#        SIMPLE CALCULATOR
#   ============================
#   1. Addition
#   2. Subtraction
#   3. Multiplication
#   4. Division
#   5. Modulus
#   6. Exponentiation
#   7. Quit
#   Select an operation (1-7):
#
# -----------------------------------------------------------------------------
# EXPECTED INTERACTION EXAMPLE
# -----------------------------------------------------------------------------
#
#   Select an operation (1-7): 4
#   Enter first number : 10
#   Enter second number: 3
#   Result: 10 / 3 = 3.33
#
#   Select an operation (1-7): 4
#   Enter first number : 5
#   Enter second number: 0
#   Error: Cannot divide by zero.
#
#   Select an operation (1-7): 7
#   Goodbye!
#
# -----------------------------------------------------------------------------
# REQUIREMENTS
# -----------------------------------------------------------------------------
# - Each arithmetic operation MUST be written as its own function.
# - Use a loop so the calculator keeps running until the user selects Quit.
# - Division by zero must be caught and handled with a clear error message
#   (do NOT let the program crash).
# - Division results should be rounded to 2 decimal places.
# - Handle invalid menu choices gracefully.
#

#
# =============================================================================
# YOUR CODE BELOW — remove the # symbols from the scaffold and fill it in
# =============================================================================


#                         SOLUTION
# - Arithmetic Functions -
def add(a, b):
    return a + b

def subtract(a, b):
    return a - b

def multiply(a, b):
    return a * b

def divide(a, b):
    if b == 0:
        return "Error: Cannot divide by zero."
    return round(a / b, 2)

def modulus(a, b):
    if b == 0:
        return "Error: Cannot divide by zero."
    return a % b

def exponentiate(a, b):
    return a ** b


def display_menu():
    print("=========================")
    print("    SIMPLE CALCULATOR    ")
    print("=========================")
    print("1. Addition")
    print("2. Subtraction")
    print("3. Multiplication")
    print("4. Division")
    print("5. Modulus")
    print("6. Exponentiation")
    print("7. Quit")


def main():
    while True:
        display_menu()
        choice = input("Select an operation (1-7): ").strip()

        if choice == "7":
            print("Goodbye!")
            break

        if choice not in ["1", "2", "3", "4", "5", "6"]:
            print("Invalid selection. Please choose a number from 1 to 7.\n")
            continue

        try:
            num1 = float(input("Enter first number : "))
            num2 = float(input("Enter second number: "))
        except ValueError:
            print("Error: Invalid number input.\n")
            continue

        # Format number output cleanly (show whole numbers as integers)
        fmt1 = int(num1) if num1.is_integer() else num1
        fmt2 = int(num2) if num2.is_integer() else num2

        if choice == "1":
            res = add(num1, num2)
            fmt_res = int(res) if isinstance(res, float) and res.is_integer() else res
            print(f"Result: {fmt1} + {fmt2} = {fmt_res}\n")

        elif choice == "2":
            res = subtract(num1, num2)
            fmt_res = int(res) if isinstance(res, float) and res.is_integer() else res
            print(f"Result: {fmt1} - {fmt2} = {fmt_res}\n")

        elif choice == "3":
            res = multiply(num1, num2)
            fmt_res = int(res) if isinstance(res, float) and res.is_integer() else res
            print(f"Result: {fmt1} * {fmt2} = {fmt_res}\n")

        elif choice == "4":
            res = divide(num1, num2)
            if isinstance(res, str):
                print(f"{res}\n")
            else:
                fmt_res = int(res) if res.is_integer() else res
                print(f"Result: {fmt1} / {fmt2} = {fmt_res}\n")

        elif choice == "5":
            res = modulus(num1, num2)
            if isinstance(res, str):
                print(f"{res}\n")
            else:
                fmt_res = int(res) if res.is_integer() else res
                print(f"Result: {fmt1} % {fmt2} = {fmt_res}\n")

        elif choice == "6":
            res = exponentiate(num1, num2)
            fmt_res = int(res) if isinstance(res, float) and res.is_integer() else res
            print(f"Result: {fmt1} ** {fmt2} = {fmt_res}\n")


if __name__ == "__main__":
    main()

File: programming-fundamentals-asssignment-2-ofoe01-boafo/assignment_09_simple_calculator.cpp
// =============================================================================
// PROGRAMMING FUNDAMENTALS — Assignment 9
// =============================================================================
//
// TASK: Console-Based Simple Calculator
//
// Build a calculator program that runs in the console and performs basic
// arithmetic operations based on the user's input.
//
// -----------------------------------------------------------------------------
// OPERATIONS YOUR CALCULATOR MUST SUPPORT
// -----------------------------------------------------------------------------
//
//   1. Addition          ( + )    e.g.  10 + 3  =  13
//   2. Subtraction       ( - )    e.g.  10 - 3  =  7
//   3. Multiplication    ( * )    e.g.  10 * 3  =  30
//   4. Division          ( / )    e.g.  10 / 3  =  3.33
//   5. Modulus           ( % )    e.g.  10 % 3  =  1  (remainder)
//   6. Exponentiation    ( ^ )    e.g.  2 ^ 8   =  256
//   7. Quit
//
// -----------------------------------------------------------------------------
// HOW THE MENU SHOULD LOOK
// -----------------------------------------------------------------------------
//
//   ============================
//        SIMPLE CALCULATOR
//   ============================
//   1. Addition
//   2. Subtraction
//   3. Multiplication
//   4. Division
//   5. Modulus
//   6. Exponentiation
//   7. Quit
//   Select an operation (1-7):
//
// -----------------------------------------------------------------------------
// EXPECTED INTERACTION EXAMPLE
// -----------------------------------------------------------------------------
//
//   Select an operation (1-7): 4
//   Enter first number : 10
//   Enter second number: 3
//   Result: 10 / 3 = 3.33
//
//   Select an operation (1-7): ssssss
//   Enter first number : 5
//   Enter second number: 0
//   Error: Cannot divide by zero.
//
//   Select an operation (1-7): 7
//   Goodbye!
//
// -----------------------------------------------------------------------------
// REQUIREMENTS
// -----------------------------------------------------------------------------
// - Each arithmetic operation MUST be written as its own function.
// - Use a loop so the calculator keeps running until the user selects Quit.
// - Division by zero must be caught and handled with a clear error message
//   (do NOT let the program crash).
// - Use fixed and setprecision(2) to display results to 2 decimal places.
// - Handle invalid menu choices gracefully.
// - For exponentiation use a loop or the pow() function from <cmath>.
//

// =============================================================================
// YOUR CODE BELOW — remove the // symbols from the scaffold and fill it in
// =============================================================================


//                        SOLUTION

#include <iostream>
#include <iomanip>
#include <cmath>
#include <limits>

// 1.
double add(double a, double b) {
    return a + b;
}

// 2.
double subtract(double a, double b) {
    return a - b;
}

// 3. 
double multiply(double a, double b) {
    return a * b;
}

// 4.
bool divide(double a, double b, double& result) {
    if (b == 0) {
        return false;
    }
    result = a / b;
    return true;
}

// 5.
bool calculateModulus(double a, double b, int& result) {
    int intA = static_cast<int>(a);
    int intB = static_cast<int>(b);
    if (intB == 0) {
        return false;
    }
    result = intA % intB;
    return true;
}

// 6.
double power(double base, double exp) {
    return std::pow(base, exp);
}

// Display of menu
void displayMenu() {
    std::cout << "=================================\n";
    std::cout << "        SIMPLE CALCULATOR        \n";
    std::cout << "=================================\n";
    std::cout << "1. Addition\n";
    std::cout << "2. Subtraction\n";
    std::cout << "3. Multiplication\n";
    std::cout << "4. Division\n";
    std::cout << "5. Modulus\n";
    std::cout << "6. Exponentiation\n";
    std::cout << "7. Quit\n";
}

int main() {
    int choice = 0;

    displayMenu();
    while (choice != 7) {
        std::cout << "\nSelect an operation (1-7): ";
        if (!(std::cin >> choice)) {
            std::cin.clear();
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            std::cout << "Invalid menu choice! Please enter a number between 1 and 7.\n";
            continue;
        }

        if (choice == 7) {
            std::cout << "Goodbye!\n";
            break;
        }

        if (choice < 1 || choice > 7) {
            std::cout << "Invalid menu choice! Please enter a number between 1 and 7.\n";
            continue;
        }

        double num1, num2;
        std::cout << "Enter first number : ";
        std::cin >> num1;
        std::cout << "Enter second number: ";
        std::cin >> num2;
        std::cout << std::fixed << std::setprecision(2);

        // Perform calculation based on user's choice
        switch (choice) {
            case 1:
                std::cout << "Result: " << num1 << " + " << num2 << " = " << add(num1, num2) << "\n";
                break;

            case 2:
                std::cout << "Result: " << num1 << " - " << num2 << " = " << subtract(num1, num2) << "\n";
                break;

            case 3:
                std::cout << "Result: " << num1 << " * " << num2 << " = " << multiply(num1, num2) << "\n";
                break;

            case 4: {
                double result;
                if (divide(num1, num2, result)) {
                    std::cout << "Result: " << num1 << " / " << num2 << " = " << result << "\n";
                } else {
                    std::cout << "Error: Cannot divide by zero.\n";
                }
                break;
            }

            case 5: {
                int result;
                if (calculateModulus(num1, num2, result)) {
                    std::cout << "Result: " << static_cast<int>(num1) << " % " << static_cast<int>(num2) << " = " << result << "\n";
                } else {
                    std::cout << "Error: Cannot divide by zero.\n";
                }
                break;
            }

            case 6:
                std::cout << "Result: " << num1 << " ^ " << num2 << " = " << power(num1, num2) << "\n";
                break;
        }
    }

    return 0;
}